plugins/outputs/influxdb/http.go

```diff
@@ -236,55 +238,66 @@ func (c *httpClient) CreateDatabase(ctx context.Context, database string) error
 	}
 }
 
+type dbrp struct {
+	Database        string
+	RetentionPolicy string
+}
+
 // Write sends the metrics to InfluxDB
 func (c *httpClient) Write(ctx context.Context, metrics []telegraf.Metric) error {
-	batches := make(map[string][]telegraf.Metric)
-	if c.config.DatabaseTag == "" {
-		err := c.writeBatch(ctx, c.config.Database, metrics)
-		if err != nil {
-			return err
+	// If these options are not used, we can skip in plugin batching and send
+	// the full batch in a single request.
+	if c.config.DatabaseTag == "" && c.config.RetentionPolicyTag == "" {
+		return c.writeBatch(ctx, c.config.Database, c.config.RetentionPolicy, metrics)
+	}
+
+	batches := make(map[dbrp][]telegraf.Metric)
+	for _, metric := range metrics {
+		db, ok := metric.GetTag(c.config.DatabaseTag)
+		if !ok {
+			db = c.config.Database
 		}
-	} else {
-		for _, metric := range metrics {
-			db, ok := metric.GetTag(c.config.DatabaseTag)
-			if !ok {
-				db = c.config.Database
-			}
 
-			if _, ok := batches[db]; !ok {
-				batches[db] = make([]telegraf.Metric, 0)
-			}
+		rp, ok := metric.GetTag(c.config.RetentionPolicyTag)
+		if !ok {
+			rp = c.config.RetentionPolicy
+		}
 
-			if c.config.ExcludeDatabaseTag {
-				// Avoid modifying the metric in case we need to retry the request.
-				metric = metric.Copy()
-				metric.Accept()
-				metric.RemoveTag(c.config.DatabaseTag)
-			}
+		dbrp := dbrp{
+			Database:        db,
+			RetentionPolicy: rp,
+		}
 
-			batches[db] = append(batches[db], metric)
+		if c.config.ExcludeDatabaseTag || c.config.ExcludeRetentionPolicyTag {
+			// Avoid modifying the metric in case we need to retry the request.
+			metric = metric.Copy()
+			metric.Accept()
+			metric.RemoveTag(c.config.DatabaseTag)
+			metric.RemoveTag(c.config.RetentionPolicyTag)
 		}
 
-		for db, batch := range batches {
-			if !c.config.SkipDatabaseCreation && !c.createdDatabases[db] {
-				err := c.CreateDatabase(ctx, db)
-				if err != nil {
-					c.log.Warnf("When writing to [%s]: database %q creation failed: %v",
-						c.config.URL, db, err)
-				}
-			}
+		batches[dbrp] = append(batches[dbrp], metric)
+	}
 
-			err := c.writeBatch(ctx, db, batch)
+	for dbrp, batch := range batches {
+		if !c.config.SkipDatabaseCreation && !c.createdDatabases[dbrp.Database] {
+			err := c.CreateDatabase(ctx, dbrp.Database)
 			if err != nil {
-				return err
+				c.log.Warnf("When writing to [%s]: database %q creation failed: %v",
+					c.config.URL, dbrp.Database, err)
 			}
 		}
+
+		err := c.writeBatch(ctx, dbrp.Database, dbrp.RetentionPolicy, batch)
+		if err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-func (c *httpClient) writeBatch(ctx context.Context, db string, metrics []telegraf.Metric) error {
-	url, err := makeWriteURL(c.config.URL, db, c.config.RetentionPolicy, c.config.Consistency)
+func (c *httpClient) writeBatch(ctx context.Context, db, rp string, metrics []telegraf.Metric) error {
```

Comment:
rp might not be super clear as retentionPolicy

Reply:
Agreed, but I think it should be okay as it matches InfluxDB lingo https://docs.influxdata.com/influxdb/v1.7/concepts/glossary#retention-policy-rp